Dependency MQTTnet out of date and has breaking changes.

The dependency for MQTTnet is now at version 4.3.1.873, which includes some breaking changes (or at least some binary incompatibilities).

Attempting a project like so:
```csproj
<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="CloudNative.CloudEvents" Version="2.7.1" />
    <PackageReference Include="CloudNative.CloudEvents.Mqtt" Version="2.7.1" />
    <PackageReference Include="CloudNative.CloudEvents.SystemTextJson" Version="2.7.1" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="6.0.18" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="6.0.18" PrivateAssets="all" />
    <PackageReference Include="MQTTnet" Version="3.1.2" />
  </ItemGroup>

</Project>
``` 
... ends up yielding
```bash
dotnet build
MSBuild version 17.3.2+561848881 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
/workspaces/event-client/Pages/FetchData.razor(64,30): error CS0012: The type 'MqttApplicationMessage' is defined in an assembly that is not referenced. You must add a reference to assembly 'MQTTnet, Version=3.0.15.0, Culture=neutral, PublicKeyToken=b69712f52770c0a7'. [/workspaces/event-client/event-client.csproj]
/workspaces/event-client/Pages/FetchData.razor(65,37): warning CS8602: Dereference of a possibly null reference. [/workspaces/event-client/event-client.csproj]
/workspaces/event-client/Pages/FetchData.razor(65,26): warning CS8604: Possible null reference argument for parameter 'item' in 'void List<string>.Add(string item)'. [/workspaces/event-client/event-client.csproj]

Build FAILED.

/workspaces/event-client/Pages/FetchData.razor(65,37): warning CS8602: Dereference of a possibly null reference. [/workspaces/event-client/event-client.csproj]
/workspaces/event-client/Pages/FetchData.razor(65,26): warning CS8604: Possible null reference argument for parameter 'item' in 'void List<string>.Add(string item)'. [/workspaces/event-client/event-client.csproj]
/workspaces/event-client/Pages/FetchData.razor(64,30): error CS0012: The type 'MqttApplicationMessage' is defined in an assembly that is not referenced. You must add a reference to assembly 'MQTTnet, Version=3.0.15.0, Culture=neutral, PublicKeyToken=b69712f52770c0a7'. [/workspaces/event-client/event-client.csproj]
    2 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.73
```

--- a/RELEASING.md
+++ b/RELEASING.md
@@ -41,14 +41,27 @@ acts as encouragement to use the latest version of the core package.
 
 Within this repository, this is achieved by the following mechanisms:
 
-- Individual csproj files do not specify a version
+- Most individual csproj files do not specify a version
 - The [Directory.Build.props](src/Directory.Build.props) file has a `<Version>` element
-  specifying the version of all packages
+  specifying the version of all packages which don't need a separate
+  major version
 
-A single GitHub release (and tag) will be created for each beta release, to cover all packages.
+A single GitHub release (and tag) will be created for each release,
+to cover all packages.
 
 - Example tag name: "CloudNative.CloudEvents.All-2.0.0"
 - Example release title: "All packages version 2.0.0"
+
+### Exception: packages with different major versions
+
+For some "satellite" packages, we need a different major version
+number, typically to adopt a new major version of a dependency. In
+this case, the satellite package will have its own major version,
+but keep the minor and patch version of everything else.
+
+For example, in order to take a new major version of the `MQTTnet`
+dependency, `CloudNative.CloudEvents.Mqtt` 3.8.0 was released with
+version 2.8.0 of all other packages.
 
 ## New / unstable package versioning
 

--- a/src/CloudNative.CloudEvents.Mqtt/CloudNative.CloudEvents.Mqtt.csproj
+++ b/src/CloudNative.CloudEvents.Mqtt/CloudNative.CloudEvents.Mqtt.csproj
@@ -5,11 +5,14 @@
     <Description>MQTT extensions for CloudNative.CloudEvents</Description>
     <PackageTags>cncf;cloudnative;cloudevents;events;mqtt</PackageTags>
     <LangVersion>8.0</LangVersion>
+    <Version>3.$(MinorVersion).$(PatchVersion)</Version>
+    <!-- After the first release of v3, we'll change the major here to 3. -->
+    <PackageValidationBaselineVersion>2.$(PackageValidationMinor).0</PackageValidationBaselineVersion>
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="MQTTnet" Version="3.0.15" />
+    <PackageReference Include="MQTTnet" Version="4.3.6.1152" />
     <ProjectReference Include="..\CloudNative.CloudEvents\CloudNative.CloudEvents.csproj" />
   </ItemGroup>
 

--- a/src/CloudNative.CloudEvents.Mqtt/MqttExtensions.cs
+++ b/src/CloudNative.CloudEvents.Mqtt/MqttExtensions.cs
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Cloud Native Foundation.
+// Copyright (c) Cloud Native Foundation.
 // Licensed under the Apache 2.0 license.
 // See LICENSE file in the project root for full license information.
 
@@ -39,10 +39,10 @@ namespace CloudNative.CloudEvents.Mqtt
             Validation.CheckNotNull(message, nameof(message));
 
             // TODO: Determine if there's a sensible content type we should apply.
-            return formatter.DecodeStructuredModeMessage(message.Payload, contentType: null, extensionAttributes);
+            return formatter.DecodeStructuredModeMessage(message.PayloadSegment, contentType: null, extensionAttributes);
         }
 
-        // TODO: Update to a newer version of MQTTNet and support both binary and structured mode?
+        // TODO: Support both binary and structured mode.
         /// <summary>
         /// Converts a CloudEvent to <see cref="MqttApplicationMessage"/>.
         /// </summary>
@@ -61,7 +61,7 @@ namespace CloudNative.CloudEvents.Mqtt
                     return new MqttApplicationMessage
                     {
                         Topic = topic,
-                        Payload = BinaryDataUtilities.AsArray(formatter.EncodeStructuredModeMessage(cloudEvent, out _))
+                        PayloadSegment = BinaryDataUtilities.GetArraySegment(formatter.EncodeStructuredModeMessage(cloudEvent, out _))
                     };
                 default:
                     throw new ArgumentOutOfRangeException(nameof(contentMode), $"Unsupported content mode: {contentMode}");

--- a/src/CloudNative.CloudEvents/Core/BinaryDataUtilities.cs
+++ b/src/CloudNative.CloudEvents/Core/BinaryDataUtilities.cs
@@ -114,7 +114,16 @@ namespace CloudNative.CloudEvents.Core
         }
 
         // Note: when this returns, the Array property of the returned segment is guaranteed not to be null.
-        private static ArraySegment<byte> GetArraySegment(ReadOnlyMemory<byte> memory) =>
+
+        /// <summary>
+        /// Returns the data from <paramref name="memory"/> as a byte array, return the underlying array
+        /// if there is one, or creating a copy otherwise. This method should be used with care, due to the
+        /// "sometimes shared, sometimes not" nature of the result. (It is generally safe to use this with the result
+        /// of encoding a CloudEvent, assuming the same memory is not used elsewhere.)
+        /// </summary>
+        /// <param name="memory">The memory to obtain the data from.</param>
+        /// <returns>The data in <paramref name="memory"/> as an array segment.</returns>
+        public static ArraySegment<byte> GetArraySegment(ReadOnlyMemory<byte> memory) =>
             MemoryMarshal.TryGetArray(memory, out var segment) && segment.Array is not null
                 ? segment
                 : new ArraySegment<byte>(memory.ToArray());

--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -1,11 +1,22 @@
 <Project>
   <PropertyGroup>
     <!-- 
-      - We use the same version number for all stable
-      - packages. See RELEASING.md for details.
+      - We use the same minor/patch version number for all stable
+      - packages, and the same major for most packages.
+      - See RELEASING.md for details.
       -->
-    <Version>2.7.1</Version>
-    <PackageValidationBaselineVersion>2.7.0</PackageValidationBaselineVersion>
+    <MajorVersion>2</MajorVersion>
+    <MinorVersion>7</MinorVersion>
+    <PatchVersion>1</PatchVersion>
+    <PackageValidationMinor>7</PackageValidationMinor>
+    <Version>$(MajorVersion).$(MinorVersion).$(PatchVersion)</Version>
+    <!-- 
+      - The version used for detecting breaking changes.
+      - This is always older than the current version (except when creating a new major)
+      - and is the patch-0 of either the current minor (if the current patch is non-zero)
+      - or the previous minor (if the current patch is zero).
+      -->
+    <PackageValidationBaselineVersion>2.$(PackageValidationMinor).0</PackageValidationBaselineVersion>
     
     <!-- Make the repository root available for other properties -->
     <RepoRoot>$([System.IO.Path]::GetDirectoryName($([MSBuild]::GetPathOfFileAbove('.gitignore', '$(MSBuildThisFileDirectory)'))))</RepoRoot>

--- a/test/CloudNative.CloudEvents.UnitTests/Mqtt/MqttTest.cs
+++ b/test/CloudNative.CloudEvents.UnitTests/Mqtt/MqttTest.cs
@@ -1,12 +1,10 @@
-﻿// Copyright (c) Cloud Native Foundation.
+// Copyright (c) Cloud Native Foundation.
 // Licensed under the Apache 2.0 license.
 // See LICENSE file in the project root for full license information.
 
 using CloudNative.CloudEvents.NewtonsoftJson;
 using MQTTnet;
 using MQTTnet.Client;
-using MQTTnet.Client.Options;
-using MQTTnet.Client.Receiving;
 using MQTTnet.Server;
 using System;
 using System.Net.Mime;
@@ -18,16 +16,17 @@ namespace CloudNative.CloudEvents.Mqtt.UnitTests
 {
     public class MqttTest : IDisposable
     {
-        private readonly IMqttServer mqttServer;
+        private readonly MqttServer mqttServer;
 
         public MqttTest()
         {
             var optionsBuilder = new MqttServerOptionsBuilder()
                 .WithConnectionBacklog(100)
+                .WithDefaultEndpoint()
                 .WithDefaultEndpointPort(52355);
 
-            this.mqttServer = new MqttFactory().CreateMqttServer();
-            mqttServer.StartAsync(optionsBuilder.Build()).GetAwaiter().GetResult();
+            this.mqttServer = new MqttFactory().CreateMqttServer(optionsBuilder.Build());
+            mqttServer.StartAsync().GetAwaiter().GetResult();
         }
 
         public void Dispose()
@@ -55,14 +54,17 @@ namespace CloudNative.CloudEvents.Mqtt.UnitTests
 
             var options = new MqttClientOptionsBuilder()
                 .WithClientId("Client1")
-                .WithTcpServer("localhost", 52355)
+                .WithTcpServer("127.0.0.1", 52355)
                 .WithCleanSession()
                 .Build();
 
             TaskCompletionSource<CloudEvent> tcs = new TaskCompletionSource<CloudEvent>();
             await client.ConnectAsync(options);
-            client.ApplicationMessageReceivedHandler = new MqttApplicationMessageReceivedHandlerDelegate(
-                args => tcs.SetResult(args.ApplicationMessage.ToCloudEvent(jsonEventFormatter)));
+            client.ApplicationMessageReceivedAsync += args =>
+            {
+                tcs.SetResult(args.ApplicationMessage.ToCloudEvent(jsonEventFormatter));
+                return Task.CompletedTask;
+            };
 
             var result = await client.SubscribeAsync("abc");
             await client.PublishAsync(cloudEvent.ToMqttApplicationMessage(ContentMode.Structured, new JsonEventFormatter(), topic: "abc"));

